Add GET endpoint to view details in a single record

LBHBomotoso opened the pull request:
Adds an endpoint to view the details within a single Resident Information object. 
The object is retrieved by its `academyId` (which would refer to a composite key made from of a `claim_id` value and a `person_ref` value).

**Any suggestions or comments on this PR would be appreciated.**

- [x] Return NI Number as part of the record.
- [x] Have `academyId` refer to a composite key of `claim_id` and `person_ref`
- [x] Have the ID attribute present in the Resident Information Object.
- [x] Edit the URI for the getbyid appropriately.

LBHBReynolds-Carr committed:
Terraform and Lambda Deployment

- Updated terraform file to call for aws parameters and set up
docker infrastructure
- Added deploy to development workflow

LBHBReynolds-Carr committed:
Deployment via CI/CD Pipeline

- Add persisting workspace to CI Config
- Applied environment to all terraform init commands
- Brought terraform files into line with CI requirements
- Extracted assume role job contents to assume role and persist workspace command
- Add lambda entrypoint (as per recent update to base-api)
- Removed up test-database command from docker-compose as this happens automatically with configuration

LBHBomotoso committed:
Get resident by id usecase and controller test

elcorbs commented on AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:
```diff
@@ -1,26 +1,25 @@
 
     <IsPackable>false</IsPackable>
     <IsTestProject>true</IsTestProject>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```
I think this is the code analysis that @Renny4Real added, so don't think it should be deleted

LBHBomotoso replied:
It doesn't compile with the settings

elcorbs commented on AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:
```diff
@@ -1,26 +1,25 @@
-    <PackageReference Include="Moq" Version="4.10.1" />
-    <PackageReference Include="nunit" Version="3.11.0" />
-    <PackageReference Include="Npgsql.EntityFrameworkCore.PostgreSQL" Version="3.1.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.4" />
```
We aren't using SqlServer so should be able to delete this

LBHBomotoso replied:
This was in the original project.

elcorbs commented on AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:
```diff
@@ -1,26 +1,25 @@
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+    <PackageReference Include="Bogus" Version="29.0.2" />
```
Also don't think we are using Bogus (I realise these were already there but good to tidy up anyway :))

LBHBomotoso replied:
This is in the original project

louisebolden replied:
Do you mean the base-api project? Some things are still included there as options, which should be removed from templated/descendent projects if they do not need to use it.

Perhaps the `SqlServer` and `Bogus` packages here are examples of that, and should therefore indeed be removed here? (Certainly it seems like `SqlServer` should be.)

elcorbs commented on AcademyResidentInformationApi/AcademyResidentInformationApi.csproj:
```diff
@@ -6,22 +6,18 @@
     <PackageReference Include="Microsoft.AspNetCore.Mvc.Versioning.ApiExplorer" Version="4.1.1" />
-    <PackageReference Include="Microsoft.CodeAnalysis.FxCopAnalyzers" Version="3.0.0" />
-    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.3" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="3.1.4" />
```
Again - we shouldn't need this anymore

LBHBomotoso replied:
Updated as it is in the original project

elcorbs commented on AcademyResidentInformationApi/V1/UseCase/Interfaces/IGetResidentByIdUseCase.cs:
```diff
@@ -0,0 +1,9 @@
+{
+    public interface IGetResidentByIdUseCase
+    {
+        public ResidentInformation Execute(int id);
```
You don't need the `public` accessor in interfaces as everything is public

LBHBomotoso replied:
true, same for the interface declaration?

elcorbs replied:
I think you still need it there

LBHBomotoso committed:
Update AcademyResidentInformationApi.csproj

LBHBomotoso committed:
get resident using academy ID
End to end tests and gateway tests
Fix imports ordering

Renny4Real committed:
Add gateway method to retrieve and return a single record

Renny4Real committed:
Add Get-Resident-By-Id Usecase and relevant tests

Renny4Real committed:
Add endpoint for viewing all the details within a single record for a specific id

Renny4Real committed:
Add NINumber attribute to the Resident Information Object

Renny4Real committed:
Set up test database to reflect composite primary key on HBMember and composite foreign key on HBHousehold.

The person table has a primary key that comprised of the "Claim ID" and "Person Ref".
The address table links to the person table by a foreign key that references a unique multiple column constraint on the "Claim ID" and "House ID" columns.

Renny4Real committed:
Add Academy ID which refers to the PK of the Person table.

The primary key of the person table is a composite key made up of the claim_id and the person_ref.
This commit adds an Academy ID attribute to the response object from the view record endpoint.

The Academy ID is a string of the form "{claim_id}-{person ref}".
This ID is deconstructed when a request is made and the two component values ('ClaimID' and 'PersonRef') are used to find the relevant record.

humulla reviewed (commented):
I think the overall changes made are great, I would agree to renaming residents to be claimant to make it more clear

elcorbs commented on AcademyResidentInformationApi.Tests/AcademyResidentInformationApi.Tests.csproj:
```diff
@@ -1,26 +1,25 @@
 
     <IsPackable>false</IsPackable>
     <IsTestProject>true</IsTestProject>
-    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
```
Does this still need to be added back in?

Renny4Real replied:
Thanks for spotting this. I missed this while I was rebasing. 
Will add this back in.

Renny4Real replied:
Have added this in 86dd418

elcorbs commented on AcademyResidentInformationApi/V1/Infrastructure/AcademyContext.cs:
```diff
@@ -15,15 +15,35 @@ public AcademyContext(DbContextOptions options) : base(options)
+
+            // composite foreign key for Address targeting unique columns in the Person table
+            modelBuilder.Entity<Person>()
+                .HasOne(x => x.Address)
```
Is there definitely only one address record for each person?

Renny4Real replied:
I think it's one-to-one.
From what I remember, the Address and Person tables both have `HouseID` and `ClaimID` columns and you can join the tables on these columns e.g. 
```
JOIN hbhousehold ON hbmember.claim_id = hbhousehold.claim_id
		AND hbmember.house_id = hbhousehold.house_id
```
But I will go and confirm this. 👍

Renny4Real committed:
Add Code Analysis Tag in Test Project File

Renny4Real commented:
On hindsight, would renaming the Resident to Claimant clash with the work currently being done with the search parameters.
Maybe the rename should be done once the search parameters' changes are finished?

humulla commented:
That's a good call - I think the renaming should be done once the search parameters task is complete

LBHBomotoso reviewed (commented):
https://github.com/LBHackney-IT/academy-resident-information-api/pull/4/commits/7fb380f42d6e2c553fd2b9dabcc60c9b6b3c2ce1#diff-2a845cfdb562e239d9c04fa9df262507

Line 35 in this commit. We had to suppress the static analysis for string culture/comparison

makeyw commented on database/schema.sql:
```diff
@@ -97,7 +58,48 @@ CREATE TABLE hbmember (
 	sanction_type int NULL,
-	sanction_rate_type int NULL
+	sanction_rate_type int NULL,
+	CONSTRAINT PK_hbmember PRIMARY KEY (claim_id, person_ref),
```
Might be getting confused but person_ref isn't a PK in the Ingres table.
PK is made up of claim_id, house_id and member_id

Renny4Real replied:
Yeah, I was told `person_ref` was part of the PK before I finding out that it wasn't in Ingres from our conversation.

makeyw replied:
in which case you're perfectly correct....let's have that meeting :)

bjpirt replied:
My bad! I thought it was but it looks like Single View is wrong

Renny4Real replied:
No worries!

Renny4Real commented:
Merging this as per team discussion so that we can progress with fleshing out the API functionality and renaming the endpoint from `/resident` to `/claimant`. See #17.
Any changes needed will be addressed in a separate PR. 😄
